### ISSUE TITLE
OAS 3.0 Support Backlog

### ISSUE THREAD

**Issue by shockey** [swagger-api/swagger-ui#3641]
This is a collection ticket for OAS3 specification features that are not yet supported by Swagger-UI.

Items that affect both UI and Client will be listed here as well, but Client-only items will be filed in the Client issue tracker.

### Rendering
- Path and operation-level servers are not displayed
- Link-level servers are not displayed
- Path-level `summary` and `description` is not applied/displayed
- Parameter `allowEmptyValue` is not displayed
- Parameter `style`/`explode` is not displayed
- Parameter `allowReserved` is not displayed
- Parameter `example`/`examples` is not displayed
- Parameter `content` is not displayed
- Media type `encoding` is not displayed
- Media type `example`/`examples` is not displayed
- Callbacks are not rendered
- Discriminator objects are not displayed

### Try-It-Out
- Links can't be used to stage another operation
- Link-level servers are not available for executing requests
- Blacklisted header names (`Accept`, `Content-Type`, `Authorization`) are not filtered out
- Path and operation-level servers are not available for executing requests
- Parameter `allowEmptyValue` is not considered in request value validation
- Parameter `style`/`explode` is not considered in request construction
  - Implementation may be best served by a split of Client's `buildRequest` based on definition version
- Security types `apiKey`, `http`, `openIdConnect` are not supported

**Comment by shockey** [comment]
cc: @webron

**Comment by hkosova** [comment]
If I may, I would add:

### Rendering
* [ ] Server `description` is not displayed
* [ ] Server variable `description` is not displayed

### Try It Out / Rendering
* [ ] `enum` not displayed as dropdowns
* [ ] missing UI to add/remove `array` items

**Comment by zhandao** [comment]
> enum not displayed as dropdowns
In the OAS3, `enum` is moved into `schema`, UI has not yet support this.

**Comment by leehicks** [comment]
Is there an update on when Authentication/Authorization will be supported by the UI? We used the "authorizations" in 2.0 heavily and are wanting to release with 3.0 support soon. This is a big missing piece for us. I saw mention in #3774, #3705, #3665 that it is being worked on in this issue number.

**Comment by webron** [comment]
Yup. The authorization should be included by in this week's release. Keep in mind it's more about compatibility with what existed in 2.0 (base auth, oauth2, and api key). That means that OIDC is _not_ going to be available yet. As for bearer (JWT!), that _might_ be in, but @shockey would be able to verify.

**Comment by shockey** [comment]
```yaml
type: http
scheme: bearer
```
will be supported 😄

**Comment by webron** [comment]
O.M.G. !!!! !! ! 😱

**Comment by shockey** [comment]
@leehicks, FYI, support for most 3.0 security schemes were added in tonight's release! Check out the release notes for more info.

**Comment by leehicks** [comment]
By release notes, do you mean something other than README or commit texts? I pulled the latest and the Auth button shows up and works, but when I try to use the "Try it", using either apiKey or basic http auth, I get a console error in the dist folder usage...

Uncaught TypeError: f.forEach is not a function
    at i (index.js:1)
    at t.default (index.js:1)
    at Object.o [as buildRequest] (index.js:1)
    at actions.js:234
    at utils.js:121
    at bindActionCreators.js:3
    at wrap-actions.js:14
    at Object.r [as executeRequest] (system.js:166)
    at actions.js:273
    at utils.js:121

Also, are there plans to support the requestInterceptor for adding headers like API keys, etc. outside using the Auth button (i.e. hardcoded)?

**Comment by webron** [comment]
@leehicks - release notes - https://github.com/swagger-api/swagger-ui/releases/tag/v3.4.0

If you encountered an issue with the implementation, please file a ticket, providing the requested details in the template.

**Comment by Doc-Strange** [comment]
@webron - Any progress on below issues?
1. Form data fields should be displayed as items instead of a JSON object.
2. File upload UI should appear when applicable

**Comment by webron** [comment]
We try to keep the list above up to date...

**Comment by craigjbass** [comment]
Hey - looks like https://github.com/swagger-api/swagger-ui/commit/d587d9c9ea794ea2a5afb59f583c3ed40c6321b6 was regressed at some point between 3.3.1 and 3.4.4. ( Callbacks are not rendered )

**Comment by webron** [comment]
@craigjbass thanks, see https://github.com/swagger-api/swagger-editor/issues/1565.

**Comment by amarel123** [comment]
> File upload UI should appear when applicable

What is the current progress? Its very useful option.

**Comment by rsidon** [comment]
Hi,
I am also interested in the status of File Upload UI (amarel123's question).
We really need this file chooser for our swagger playground.
Is there any update about the status of this issue? 
10x

**Comment by shockey** [comment]
Some updates:
- `anyOf / oneOf examples: Come up with solution-ish, or document to say its not something we're going to fix` was removed in favor of being tracked in #3803 and #2651.
- `Path and operation-level servers are not displayed` was removed since it was tracked in #3820 and implemented in #3972.

Besides that, all checkbox statuses have been updated.

If a task is still unchecked... no update for now 😄 if you're here to ask for a specific thing to get done, check to see if anyone has mentioned it, and if so 👍 their comment. If no one has mentioned it, feel free to comment about it.

We'll do our best to address things here in order of demand.

**Comment by nishantgupta91** [comment]
Hi,
What's the status of "Server description is not displayed"? I also need this. Moreover, the feature with multiple examples doesn't work. @shockey  Can you please confirm if the functionality mentioned here "https://swagger.io/docs/specification/adding-examples/" actually works? Thanks

**Comment by Fannon** [comment]
Also needing the file upload back in OpenApi 3.0 :) Is there a dedicated issue to watch!

Thank you for working on it!

**Comment by jelanwar** [comment]
Hi,
Is there any updated regarding <code>Parameter example/examples is not displayed</code>? as it is very useful option to have

**Comment by jonyeezs** [comment]
> File upload UI should appear when applicable

Is there anything we can help speeding this up?

**Comment by webron** [comment]
@jonyeezs PRs help ;)

**Comment by heldersepu** [comment]
@webron the comment from @jelanwar got me thinking, "anything" ...
maybe this is an opportunity to monetize Open Source projects?

What if people interested in speeding features/bug fixes could put a coin bounty on the issues?
That should bring developers to submit PR to collect the bounty

**Comment by akalongman** [comment]
We really need the feature described in #4345

**Comment by zeromake** [comment]
[doc](https://swagger.io/specification/#security-scheme-object)
in is `header` support
``` json
api_key:
    type: "apiKey"
    name: "api_key"
    in: "header"
```

in is `cookie` not support has error

``` json
api_key:
    type: "apiKey"
    name: "api_key"
    in: "cookie"
```
error:

``` text
Schema error at securityDefinitions['api_key'].in
should be equal to one of the allowed values
allowedValues: header, query
```

**Comment by hkosova** [comment]
@zeromake make sure your spec is `openapi: 3.0.0` and not `swagger: '2.0'`.

**Comment by malutanpetronel** [comment]
..
     *        @OA\MediaType(
     *            mediaType="multipart/form-data",
     *            @OA\Schema(
     *                type="object",
     *                @OA\Property(
     *                    property="files",
     *                    description="Choose a file to upload",
     *                    type="array",
     *                    @OA\Items(
     *                        description="file",
     *                        type="string",
     *                        format="binary",
     *                    ),
...
after this I use 
    public function upload(Request $request)
    {
        //$imageContent = $request->getContent();
        $files = $request->input('files');
        $files = explode(',', $files);
        foreach ($files as $file) {
            dd($file);
        }
    }

and get 
..characters">[object File]</span>"...

How  I am suppose to play with the [object File] above to finish the upload please ?

**Comment by lcahlander** [comment]
When is multipart file upload going to be supported.  You had this before!

**Comment by lcahlander** [comment]
> > When is multipart file upload going to be supported. You had this before!
> 
> It is supported in 2.9.2.

It is not showing up for me.  Please see my stackoverflow [https://stackoverflow.com/questions/54334438/how-do-i-specify-a-multifile-upload-in-openapi](url)

**Comment by lcahlander** [comment]
> > > When is multipart file upload going to be supported. You had this before!
> > 
> > 
> > It is supported in 2.9.2.
> 
> It is not showing up for me. Please see my stackoverflow [https://stackoverflow.com/questions/54334438/how-do-i-specify-a-multifile-upload-in-openapi](url)

The picker shows up, but the UI does not appear to send the filename or mime type along with the binary.

**Comment by beltschatsar** [comment]
Hi,

Can we have status about this missing feature ?

>  ### Try-it-out

- [ ] Links can't be used to stage another operation



So that an api call output can serve as input for another call (as explained there: https://swagger.io/docs/specification/links/)

Thanks!

**Comment by J-Ibarra** [comment]
Hi there; I am trying to send multiple files in one property, as described in the [Specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#considerations-for-file-uploads), I declare

```
 /testmultipart:
    post:
      summary: Send multipart request.
      requestBody:
        content:
          multipart/form-data:
            schema:
              properties:
              # The property name 'file' will be used for all files.
                file:
                  type: array
                  items:
                    type: string
                    format: binary
```

when I send the request I spect

```curl  -X POST -H "Content-Type: multipart/form-data"  "http://localhost:3000/testmultipart" --form 'file=@file1.ext' --form 'file=@file3.ext' --form 'file=@file2.ext' '```

but I see this in the body

`{ "file": [ "string" ] }`

**Comment by silkfire** [comment]
What's the status of

- [ ] Links can't be used to stage another operation

and rendering links as hyperlinks?

**Comment by paul-uz** [comment]
Any update on Links?
